ethers-providers/src/rpc/transports/ws/backend.rs

```diff
@@ -0,0 +1,194 @@
+        trace!(text = t, "Received message");
+        match serde_json::from_str(&t) {
+            Ok(item) => {
+                trace!(%item, "Deserialized message");
+                let res = self.handler.unbounded_send(item);
+                if res.is_err() {
+                    return Err(WsClientError::DeadChannel)
+                }
+            }
+            Err(e) => {
+                error!(e = %e, "Failed to deserialize message");
+            }
+        }
+        Ok(())
+    }
+
+    #[cfg(not(target_arch = "wasm32"))]
+    async fn handle(&mut self, item: WsStreamItem) -> Result<(), WsClientError> {
+        match item {
+            Ok(item) => match item {
+                Message::Text(t) => self.handle_text(t).await,
+                Message::Ping(data) => {
+                    if self.server.send(Message::Pong(data)).await.is_err() {
+                        return Err(WsClientError::UnexpectedClose)
+                    }
+                    Ok(())
+                }
+
+                Message::Pong(_) => Ok(()),
+                Message::Frame(_) => Ok(()),
+
+                Message::Binary(buf) => Err(WsClientError::UnexpectedBinary(buf)),
+                Message::Close(frame) => {
+                    if frame.is_some() {
+                        error!("Close frame: {}", frame.unwrap());
+                    }
+                    Err(WsClientError::UnexpectedClose)
+                }
+            },
+            Err(e) => {
+                error!(err = %e, "Error response from WS");
+                Err(e.into())
+            }
+        }
+    }
+
+    #[cfg(target_arch = "wasm32")]
+    async fn handle(&mut self, item: WsStreamItem) -> Result<(), WsClientError> {
+        match item {
+            Message::Text(inner) => self.handle_text(inner).await,
+            Message::Binary(buf) => Err(WsClientError::UnexpectedBinary(buf)),
+        }
+    }
+
+    pub fn spawn(mut self) {
+        let fut = async move {
+            let mut err = false;
+            loop {
+                #[cfg(not(target_arch = "wasm32"))]
+                let keepalive = tokio::time::sleep(std::time::Duration::from_secs(10)).fuse();
+                #[cfg(not(target_arch = "wasm32"))]
+                tokio::pin!(keepalive);
+
+                // in wasm, we don't ping. as ping doesn't exist in our wasm lib
+                #[cfg(target_arch = "wasm32")]
+                let mut keepalive = futures_util::future::pending::<()>().fuse();
+
+                select! {
+                    _ = keepalive => {
+                        #[cfg(not(target_arch = "wasm32"))]
+                        if let Err(e) = self.server.send(Message::Ping(vec![])).await {
+                            error!("WS connection error {e}");
+                            err = true;
+                            break
+                        }
+                        #[cfg(target_arch = "wasm32")]
+                        unreachable!();
```

Comment:
ah nice; smart. so the future never completes so obvi it won't reach that branch

cc @ncitron in case you're using the WS code in WASM - would love you to give this new WS code a try

Reply:
the downside is that there's no keepalive on wasm ws connections. wasm doesn't support ping/pong so if we were to ever add a JS setTimeout based keepalive here, it would have to issue a real rpc request

Reply:
yup